collie/interactions/datasets.py

```diff
@@ -383,66 +417,76 @@ def __getitem__(self, index: Union[int, Iterable[int]]) -> (
         item_id = self.mat.col[index]
         # rating = self.mat.data[index]  # not needed for any loss currently implemented
 
-        negative_item_ids_array = self._negative_sample(user_id)
+        if self.negative_sample_type == 'item':
+            negative_ids_array = self._negative_sample(user_id)
+        elif self.negative_sample_type == 'user':
+            negative_ids_array = self._negative_sample(item_id)
 
-        return (user_id, item_id), negative_item_ids_array
+        return (user_id, item_id), negative_ids_array
+
+    def _negative_sample(self, id: Union[int, np.array]) -> np.array:
+        """Generate negative samples for an ``id``."""
+        if self.negative_sample_type == 'item':
+            num_ids = self.num_items
+            positive_set_tuple = '(specific_id, negative_id)'
+        elif self.negative_sample_type == 'user':
+            num_ids = self.num_users
+            positive_set_tuple = '(negative_id, specific_id)'
 
-    def _negative_sample(self, user_id: Union[int, np.array]) -> np.array:
-        """Generate negative samples for a ``user_id``."""
         if self.max_number_of_samples_to_consider > 0:
             # if we are here, we are doing true negative sampling
-            negative_item_ids_list = list()
+            negative_ids_list = list()
 
-            if not isinstance(user_id, collections.abc.Iterable):
-                user_id = [user_id]
+            if not isinstance(id, collections.abc.Iterable):
+                id = [id]
 
-            for specific_user_id in user_id:
-                # generate true negative samples for the ``user_id``
+            for specific_id in id:
+                # generate true negative samples for the ``id``
                 samples_checked = 0
-                temp_negative_item_ids_list = list()
+                temp_negative_ids_list = list()
 
-                while len(temp_negative_item_ids_list) < self.num_negative_samples:
-                    negative_item_id = random.choice(range(self.num_items))
-                    # we have a negative sample, make sure the user has not interacted with it
+                while len(temp_negative_ids_list) < self.num_negative_samples:
+                    negative_id = random.choice(range(num_ids))
+                    # we have a negative sample, make sure the seed ID has not interacted with it
                     # before, else we resample and try again
                     while (
-                        (specific_user_id, negative_item_id) in self.positive_items
-                        or negative_item_id in temp_negative_item_ids_list
+                        eval(positive_set_tuple) in self.positive_sets
```

Comment:
This is me being so nitpicky, but I generally try to stay away from `eval`. 

https://stackoverflow.com/questions/1832940/why-is-using-eval-a-bad-practice 

Is there another way we can do this without `eval`?

Reply:
💀